pkg/operator/ceph/cluster/osd/replace.go

```diff
@@ -0,0 +1,217 @@
+		}
+
+		_, err = context.Clientset.CoreV1().ConfigMaps(clusterInfo.Namespace).Create(clusterInfo.Context, newConfigMap, metav1.CreateOptions{})
+		if err != nil {
+			return errors.Wrapf(err, "failed to create osd replace configMap %q", newConfigMap.Name)
+		}
+
+		return nil
+	}
+
+	// Update existing configMap
+	existingCM.Data[OSDReplaceConfigKey] = configStr
+	if _, err := context.Clientset.CoreV1().ConfigMaps(clusterInfo.Namespace).Update(clusterInfo.Context, existingCM, metav1.UpdateOptions{}); err != nil {
+		return errors.Wrapf(err, "failed to update existing OSD replace config map %q", existingCM.Name)
+	}
+
+	return nil
+}
+
+func (o *OSDReplaceConfig) deleteCM(context *clusterd.Context, clusterInfo *cephclient.ClusterInfo) error {
+	err := context.Clientset.CoreV1().ConfigMaps(clusterInfo.Namespace).Delete(clusterInfo.Context, OSDReplaceConfigName, metav1.DeleteOptions{})
+	if err != nil && !kerrors.IsNotFound(err) {
+		return errors.Wrapf(err, "failed to delete the osd replace status config map %q", OSDReplaceConfigName)
+	}
+
+	return nil
+}
+
+func (o *OSDReplaceConfig) string() (string, error) {
+	configInBytes, err := json.Marshal(o)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to marshal osd replace config")
+	}
+
+	return string(configInBytes), nil
+}
+
+func (o *OSDReplaceConfig) doneReplacing() bool {
+	if o != nil && len(o.OSDs) == 0 {
+		return true
+	}
+	return false
+}
+
+// getNextOSDToReplace returns an OSD to be replaced if the PGs are healthy
+func (o *OSDReplaceConfig) getNextOSDToReplace() *OSDReplaceInfo {
+	// If OSDInProgress is still available in the config, then last reconcile might have failed to delete this OSD.
+	// So continue with the same OSD.
+	if o.OSDInProgress != nil {
+		return o.OSDInProgress
+	}
+
+	if len(o.OSDs) == 0 {
+		return nil
+	}
+
+	osdToReplace := &o.OSDs[0]
+	o.OSDs = o.OSDs[1:]
+	o.OSDInProgress = osdToReplace
+
+	return osdToReplace
+}
+
+func (c *Cluster) getOSDReplaceConfig() (*OSDReplaceConfig, error) {
+	replaceConfig, err := GetOSDReplaceConfigMap(c.context, c.clusterInfo)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get any existing OSD replace config")
+	}
+
+	if replaceConfig == nil {
+		logger.Info("creating a new osd replace config")
+		replaceConfig = newOSDReplaceConfig()
+	}
+
+	osdsToReplace, err := c.getOSDWithNonMatchingStore()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to list out OSDs with non matching backend store")
+	}
+	replaceConfig.OSDs = osdsToReplace
+
+	return replaceConfig, nil
+}
+
+// getOSDWithNonMatchingStore returns OSDs with osd-store label different from expected store in cephCluster spec
+func (c *Cluster) getOSDWithNonMatchingStore() ([]OSDReplaceInfo, error) {
+	listOpts := metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", k8sutil.AppAttr, AppName)}
+	deployments, err := c.context.Clientset.AppsV1().Deployments(c.clusterInfo.Namespace).List(c.clusterInfo.Context, listOpts)
```

Comment:
How many times in a reconcile are we querying the list of OSD deployments? If there are many OSDs it may be best if we don't list them very often. For example, we also query them for counting the store types in status. Could we also count them here instead of the separate query?

Reply:
We are doing it twice.  First inside `configureLocalCephCluster` (to replace OSDs) , and then inside `configureCephMonitoring` to update the OSD status. 

Need to check if there can be a better way to avoid calling this twice.